### DISCUSSION TITLE
nnUNet and DynUnet

### DISCUSSION THREAD

**Discussion by Gijz33** [Project-MONAI/MONAI#3149] [Q&A]
Hi,

As part of my assignment I'm training some models using the MONAI framework.
I'm trying to understand the differences between nnUNet and DynUNet.

As far as I understand, DynUNet is the implementation of nnUNet's architecture in MONAI, but the whole pipeline is not implemented. Is this correct?

If so, how does DynUNet differ from UNet?

Many thanks for your reply.

**Comment by yiheng-wang-nv** [comment]
Hi @Gijz33 , you can refer to the readme file in: https://github.com/Project-MONAI/tutorials/tree/master/modules/dynunet_pipeline

**Reply by Gijz33** [reply]
Hi @yiheng-wang-nv, thank you for sending the readme file.

I've looked through this file before, but I understood that DynUNet only doesn't have the postprocessing step. The rest remains the same as in nnUNET (see image). Is that correct?
![nnuU-Net pipeline](https://user-images.githubusercontent.com/92758984/138174522-c5e6a0a5-1f91-459e-8326-11e9c312c79e.png)

**Reply by yiheng-wang-nv** [reply]
Hi @Gijz33 , `heuristic rules` does not covered, and data augmentation ways of dynunet pipeline also have some differences.
The `heuristic rules` helped to determine some task specific parameters, such as batch size and input patch size, for dynunet we just make use of the achieved values.

**Reply by Gijz33** [reply]
Thank you for your help @yiheng-wang-nv !

**Reply by Gijz33** [reply]
Hi @yiheng-wang-nv, 
Your answer has already helped me to understand DynUnet better. Thank you for that. However, there are still some  things I would like to know. Maybe you can help me out? I have listed my questions to add some structure:

1.) Is documentation available on how nnUNet is implemented in MONAI? I would like to know how nn-Unet's heuristic rules are used to determine the task specific parameters in DynUnet.
 
2.) nnUnet trains 3 different archictectures (2D, 3D and 3DC U-Net), it uses the best architecture or it combines 2 architectures in the final model. How does DynUnet do this?

**Reply by Gijz33** [reply]
Hi @yiheng-wang-nv,
Maybe my message got lost in all the mail traffic:) Could you help me with these questions?

**Reply by yiheng-wang-nv** [reply]
Hi @Gijz33 , sorry for late reply.
As for your questions:
1) sorry, there are no doc of how nnUNet is implemented. The task specific parameters are referred from the paper.
2) we only implement the 3D_fullres version, the reason is that it performs the best for the most of the decathlon tasks.

**Reply by Gijz33** [reply]
Hi @yiheng-wang-nv ,
No problem, thank you for your help and answers!
